File: valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
           
        req_ending_brackets = []
        bracket_mapping = {
            '(' : ')',
            '{' : '}',
            '[' : ']'
        }
        for x in s:
            if x in bracket_mapping.keys():
                req_ending_brackets.append(bracket_mapping[x])
            
            elif x in bracket_mapping.values():
                if len(req_ending_brackets) == 0:
                    return False
                elif req_ending_brackets[-1] == x:
                    del req_ending_brackets[-1]
                else:
                    return False

        if len(req_ending_brackets) == 0:
            return True
        else:
            return False
        

File: valid_parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let req_ending_brackets = [];

    const bracket_mapping = {
        '(':')',
        '{':'}',
        '[':']'
    };
    for (const x of s){
        if (bracket_mapping.hasOwnProperty(x)){
            req_ending_brackets.push(bracket_mapping[x]);
        }
        else if(Object.values(bracket_mapping).includes(x)){
            if (req_ending_brackets.length == 0){
                return false;
            }
            else if(req_ending_brackets[req_ending_brackets.length-1] == x){
                req_ending_brackets.pop();
            }
            else{
                return false;
            }
        }
    }
    if (req_ending_brackets.length == 0){
        return true;
    }
    else{
        return false;
    }
};

// Time complexity: O(n)
// Space complexity: O(n)